File: core/treasury_agent.zig
//! treasury_agent.zig — autonomous market-making agent for the NS treasury.
//!
//! Architecture goal (memory: project_omnibus_autonomous_treasury_design):
//! the NS treasury (ens.omnibus, registrar slot 5) accumulates payments
//! from name-claim TXs. Instead of those funds sitting idle (or worse,
//! being withdrawn externally), an agent thread *cannot* withdraw them
//! either — it only places limit orders on the OMNI/USDC orderbook. This
//! gives the chain permanent on-book liquidity AND makes the treasury
//! impossible to drain (the only way out is for some user to actually
//! buy or sell against the agent's orders, on-chain).
//!
//! Design choices (set in stone unless config_user explicitly overrides):
//!   - Single pair: OMNI/USDC (the most liquid; USDC is stablecoin).
//!   - Adaptive spread: levels at mid ± k×σ for k in 1..5, where σ is the
//!     rolling stddev of the last `VOL_WINDOW` trade prices. Spread auto-
//!     widens in volatile markets, tightens when calm.
//!   - 70% of treasury balance committed to grid by default (configurable).
//!     Remaining 30% acts as renewal/premium-fee reserve.
//!   - Capital allocation per side: weighted toward k=1 (most fills happen
//!     near mid). Weights: [40, 25, 15, 12, 8] across k=1..5 (sum=100).
//!   - Re-grid trigger: cancel + re-place when mid drifts by >5% of σ from
//!     the last grid centre, OR when treasury balance changes by >10%.
//!   - Cooldown: minimum BLOCK_INTERVAL blocks between full re-grids, so a
//!     volatile market doesn't thrash the orderbook.
//!
//! Single-point-of-failure note: the agent uses the registrar slot 5
//! private key, which lives in the founder's mnemonic. Phase 2 plan is
//! to migrate this to a consensus-controlled treasury (the chain itself
//! enforces the orders, no key needed). For Phase 1 (MVP) the agent
//! signs from the slot-5 wallet derivation. Single key = single risk;
//! mitigated by the agent ONLY emitting place_order / cancel_order calls
//! — never withdraw or transfer.

const std = @import("std");
const matching_mod = @import("matching_engine.zig");
const wallet_mod = @import("wallet.zig");
const transaction_mod = @import("transaction.zig");
const blockchain_mod = @import("blockchain.zig");
const registrar_mod = @import("registrar_addresses.zig");

const Order = matching_mod.Order;
const Fill = matching_mod.Fill;

// ─── Tunable constants (defaults; override via TreasuryAgent.config) ────

/// Active pair the agent makes a market on. Hardcoded to OMNI/USDC for
/// Phase 1 — only OMNI/USDC has guaranteed liquidity from claim-TX flow.
pub const PAIR_ID: u16 = 0;

/// How many levels per side. 5 = ten orders total per re-grid.
pub const DEFAULT_LEVELS_PER_SIDE: u8 = 5;

/// Capital weights per level — the closer to mid, the more SAT we commit.
/// Index = level k-1 (0 = closest to mid). Sum is 100 (percent).
pub const DEFAULT_LEVEL_WEIGHTS = [DEFAULT_LEVELS_PER_SIDE]u8{ 40, 25, 15, 12, 8 };

/// Default fraction of treasury balance pushed into grid. The rest is
/// held back as a buffer for NS renewals / premium-fee accounting.
pub const DEFAULT_GRID_ALLOC_PCT: u8 = 70;

/// Volatility window — rolling stddev of the last N trade prices on the
/// active pair. Smaller = more reactive, larger = more stable spread.
pub const VOL_WINDOW: usize = 100;

/// Minimum and maximum k×σ spread, in micro-USD, to clamp the band when
/// stddev is near-zero (cold orderbook) or absurdly large (early outliers).
pub const MIN_SPREAD_MICRO_USD: u64 = 1_000;       // 0.001 USD
pub const MAX_SPREAD_MICRO_USD: u64 = 10_000_000;  // 10 USD

/// Re-grid every N blocks at minimum. Caps thrash on volatile markets.
pub const MIN_REGRID_BLOCKS: u64 = 10;

/// Re-grid early if mid drifts by ≥ DRIFT_THRESHOLD * sigma from the last
/// grid centre. 0.5 = re-grid as soon as mid moves by half a stddev.
pub const DRIFT_THRESHOLD_PCT: u8 = 50;

/// Re-grid early if treasury balance changes by ≥ this % from the last grid.
pub const BALANCE_DELTA_PCT: u8 = 10;

// ─── Config ────────────────────────────────────────────────────────────

pub const Config = struct {
    grid_alloc_pct: u8 = DEFAULT_GRID_ALLOC_PCT,
    levels_per_side: u8 = DEFAULT_LEVELS_PER_SIDE,
    level_weights: [DEFAULT_LEVELS_PER_SIDE]u8 = DEFAULT_LEVEL_WEIGHTS,
    min_regrid_blocks: u64 = MIN_REGRID_BLOCKS,
    drift_threshold_pct: u8 = DRIFT_THRESHOLD_PCT,
    balance_delta_pct: u8 = BALANCE_DELTA_PCT,
    vol_window: usize = VOL_WINDOW,
    enabled: bool = true,
};

// ─── Volatility tracker (rolling stddev) ────────────────────────────────

/// Ring-buffer of trade prices used to compute σ. Fixed size (VOL_WINDOW)
/// so no allocator needed after init.
pub const VolTracker = struct {
    samples: [VOL_WINDOW]u64 = [_]u64{0} ** VOL_WINDOW,
    head: usize = 0,
    count: usize = 0,

    pub fn push(self: *VolTracker, price_micro_usd: u64) void {
        self.samples[self.head] = price_micro_usd;
        self.head = (self.head + 1) % VOL_WINDOW;
        if (self.count < VOL_WINDOW) self.count += 1;
    }

    pub fn mean(self: *const VolTracker) u64 {
        if (self.count == 0) return 0;
        var sum: u128 = 0;
        for (self.samples[0..self.count]) |s| sum += s;
        return @intCast(sum / self.count);
    }

    /// Population stddev. u64 in micro-USD.
    pub fn sigma(self: *const VolTracker) u64 {
        if (self.count < 2) return 0;
        const m: i128 = @intCast(self.mean());
        var sum_sq: u128 = 0;
        for (self.samples[0..self.count]) |s| {
            const diff: i128 = @as(i128, @intCast(s)) - m;
            sum_sq += @intCast(diff * diff);
        }
        const variance = sum_sq / self.count;
        return @intCast(std.math.sqrt(@as(u128, variance)));
    }
};

// ─── Treasury agent state ───────────────────────────────────────────────

/// Tracks one of the agent's live orders so we can cancel cleanly.
const ManagedOrder = struct {
    order_id: u64,
    side: matching_mod.Side,
    level: u8,           // 1..levels_per_side
    price_micro_usd: u64,
    amount_sat: u64,
};

pub const TreasuryAgent = struct {
    /// Mostly constants set at init.
    config: Config,
    /// Treasury wallet — derived from founder mnemonic at slot 5 (.ens).
    /// Holds the secp256k1 key the agent signs every order with.
    wallet: wallet_mod.Wallet,
    /// Treasury bech32 address, cached (slice into the wallet's addr buf).
    treasury_address: []const u8,

    /// References to the systems the agent reads / writes.
    bc: *blockchain_mod.Blockchain,
    engine: *matching_mod.MatchingEngine,

    /// Volatility tracker — fed by `onFill` from rpc_server.
    vol: VolTracker = .{},

    /// Live orders we placed, indexed by slot. Capacity = 2 * levels_per_side
    /// (buys + sells). Unused slots have order_id == 0.
    live_orders: [2 * DEFAULT_LEVELS_PER_SIDE]ManagedOrder =
        [_]ManagedOrder{.{
            .order_id = 0,
            .side = .buy,
            .level = 0,
            .price_micro_usd = 0,
            .amount_sat = 0,
        }} ** (2 * DEFAULT_LEVELS_PER_SIDE),

    /// Grid centre of the last successful re-grid; used for drift check.
    last_grid_mid_micro_usd: u64 = 0,
    /// Treasury balance at the last re-grid; used for balance-delta check.
    last_grid_balance_sat: u64 = 0,
    /// Block height of the last re-grid; cooldown gate.
    last_regrid_block: u64 = 0,

    pub fn init(
        bc: *blockchain_mod.Blockchain,
        engine: *matching_mod.MatchingEngine,
        wallet: wallet_mod.Wallet,
        config: Config,
    ) TreasuryAgent {
        return .{
            .config = config,
            .wallet = wallet,
            .treasury_address = wallet.address,
            .bc = bc,
            .engine = engine,
        };
    }

    /// Called from rpc_server.exchange dispatch whenever a real-mode fill
    /// lands on the active pair. We feed the price into the volatility
    /// tracker so the next re-grid widens/tightens accordingly.
    pub fn onFill(self: *TreasuryAgent, fill: Fill) void {
        if (fill.pair_id != PAIR_ID) return;
        self.vol.push(fill.price_micro_usd);
    }

    /// Called from the mining loop after every block. Decides whether to
    /// re-grid based on cooldown + drift + balance-delta. Cheap when
    /// nothing has changed.
    pub fn tick(self: *TreasuryAgent, current_block: u64) void {
        if (!self.config.enabled) return;
        if (current_block < self.last_regrid_block + self.config.min_regrid_blocks) return;

        const balance = self.bc.getAddressBalance(self.treasury_address);
        if (balance == 0) return; // nothing to grid

        const mid = self.computeMid() orelse return; // need an orderbook ref price

        const drifted = self.midDrifted(mid);
        const balance_changed = self.balanceChanged(balance);
        const first_run = self.last_regrid_block == 0;
        if (!first_run and !drifted and !balance_changed) return;

        self.regrid(current_block, mid, balance) catch |err| {
            std.debug.print("[TREASURY-AGENT] regrid failed: {}\n", .{err});
        };
    }

    /// Compute the reference mid price. If both sides of the orderbook are
    /// populated we use (best_bid + best_ask) / 2. Otherwise we fall back
    /// to the mean of the volatility tracker (last 100 fills) so a cold
    /// book bootstraps from recent trade history. Returns null only if
    /// neither source has data yet — agent waits.
    fn computeMid(self: *const TreasuryAgent) ?u64 {
        const best_bid = self.engine.bestBid(PAIR_ID) orelse 0;
        const best_ask = self.engine.bestAsk(PAIR_ID) orelse 0;
        if (best_bid > 0 and best_ask > 0) {
            return (best_bid + best_ask) / 2;
        }
        const m = self.vol.mean();
        if (m > 0) return m;
        return null;
    }

    fn midDrifted(self: *const TreasuryAgent, current_mid: u64) bool {
        if (self.last_grid_mid_micro_usd == 0) return true;
        const sigma = @max(self.vol.sigma(), MIN_SPREAD_MICRO_USD);
        const threshold = sigma * self.config.drift_threshold_pct / 100;
        const diff = if (current_mid > self.last_grid_mid_micro_usd)
            current_mid - self.last_grid_mid_micro_usd
        else
            self.last_grid_mid_micro_usd - current_mid;
        return diff >= threshold;
    }

    fn balanceChanged(self: *const TreasuryAgent, current_balance: u64) bool {
        if (self.last_grid_balance_sat == 0) return true;
        const diff = if (current_balance > self.last_grid_balance_sat)
            current_balance - self.last_grid_balance_sat
        else
            self.last_grid_balance_sat - current_balance;
        const threshold = self.last_grid_balance_sat * self.config.balance_delta_pct / 100;
        return diff >= threshold;
    }

    /// Cancel every previously-placed live order, then place a fresh
    /// adaptive grid centred on `mid`. Capital = balance × grid_alloc_pct.
    fn regrid(self: *TreasuryAgent, current_block: u64, mid: u64, balance: u64) !void {
        // 1. Cancel everything we still own on the engine.
        for (&self.live_orders) |*mo| {
            if (mo.order_id == 0) continue;
            self.engine.cancelOrder(mo.order_id) catch {};
            mo.order_id = 0;
        }

        const sigma_clamped = std.math.clamp(self.vol.sigma(), MIN_SPREAD_MICRO_USD, MAX_SPREAD_MICRO_USD);
        const grid_capital = balance * self.config.grid_alloc_pct / 100;
        // Half the capital each side — buys consume QUOTE balance,
        // sells consume BASE. We model both as same SAT pool for now;
        // when balances diverge the engine simply rejects the side that
        // can't be funded and we re-grid on next balance change.
        const per_side_capital = grid_capital / 2;

        var placed: usize = 0;

        // 2. Place buy levels: mid - k×σ (k = 1..levels)
        var k: u8 = 1;
        while (k <= self.config.levels_per_side) : (k += 1) {
            const offset = sigma_clamped * k;
            if (offset >= mid) break; // would cross zero, skip
            const price = mid - offset;
            const weight = self.config.level_weights[k - 1];
            const level_capital_quote = per_side_capital * weight / 100;
            // amount_sat = capital_quote × 1e9 / price_micro_usd (u128 to avoid overflow)
            const amt: u64 = @intCast((@as(u128, level_capital_quote) * 1_000_000_000) / @max(price, 1));
            if (amt == 0) continue;

            const oid = self.placeOrder(.buy, price, amt) catch continue;
            self.recordLiveOrder(.buy, k, price, amt, oid, &placed);
        }

        // 3. Place sell levels: mid + k×σ
        k = 1;
        while (k <= self.config.levels_per_side) : (k += 1) {
            const offset = sigma_clamped * k;
            const price = mid + offset;
            const weight = self.config.level_weights[k - 1];
            const level_capital_base = per_side_capital * weight / 100;
            // For sells, level_capital is in BASE (OMNI SAT) directly.
            const amt = level_capital_base;
            if (amt == 0) continue;

            const oid = self.placeOrder(.sell, price, amt) catch continue;
            self.recordLiveOrder(.sell, k, price, amt, oid, &placed);
        }

        self.last_grid_mid_micro_usd = mid;
        self.last_grid_balance_sat = balance;
        self.last_regrid_block = current_block;

        std.debug.print(
            "[TREASURY-AGENT] regrid block={d} mid={d} sigma={d} placed={d}\n",
            .{ current_block, mid, sigma_clamped, placed },
        );
    }

    fn recordLiveOrder(
        self: *TreasuryAgent,
        side: matching_mod.Side,
        level: u8,
        price: u64,
        amount: u64,
        oid: u64,
        placed: *usize,
    ) void {
        for (&self.live_orders) |*mo| {
            if (mo.order_id == 0) {
                mo.* = .{
                    .order_id = oid,
                    .side = side,
                    .level = level,
                    .price_micro_usd = price,
                    .amount_sat = amount,
                };
                placed.* += 1;
                return;
            }
        }
    }

    /// Wrap matching_engine.placeOrder with the treasury wallet identity
    /// and a fresh sequential order_id. The engine accepts the order
    /// without re-verifying signature when called via this internal path
    /// (treasury owns the engine; the chain validators check the engine's
    /// trade log against the chain's UTXO set on settlement).
    fn placeOrder(
        self: *TreasuryAgent,
        side: matching_mod.Side,
        price_micro_usd: u64,
        amount_sat: u64,
    ) !u64 {
        const oid = std.time.nanoTimestamp() & 0x7FFF_FFFF_FFFF_FFFF;
        var ord = Order.empty();
        ord.order_id = @intCast(oid);
        ord.pair_id = PAIR_ID;
        ord.side = side;
        ord.price_micro_usd = price_micro_usd;
        ord.amount_sat = amount_sat;
        ord.timestamp_ms = std.time.milliTimestamp();
        ord.status = .active;

        const addr_bytes = self.treasury_address;
        const addr_len: u8 = @intCast(@min(addr_bytes.len, 64));
        @memcpy(ord.trader_address[0..addr_len], addr_bytes[0..addr_len]);
        ord.trader_addr_len = addr_len;

        try self.engine.placeOrder(ord);
        return ord.order_id;
    }
};

// ─── Tests ──────────────────────────────────────────────────────────────

const testing = std.testing;

test "VolTracker — empty has zero mean and sigma" {
    var v = VolTracker{};
    try testing.expectEqual(@as(u64, 0), v.mean());
    try testing.expectEqual(@as(u64, 0), v.sigma());
}

test "VolTracker — single sample stddev is zero" {
    var v = VolTracker{};
    v.push(1_000_000);
    try testing.expectEqual(@as(u64, 1_000_000), v.mean());
    try testing.expectEqual(@as(u64, 0), v.sigma());
}

test "VolTracker — known-population stddev matches" {
    // 1, 2, 3, 4, 5 — pop mean=3, pop variance=2, pop stddev=√2 ≈ 1
    var v = VolTracker{};
    for ([_]u64{ 1, 2, 3, 4, 5 }) |s| v.push(s);
    try testing.expectEqual(@as(u64, 3), v.mean());
    // sqrt(2) integer-truncated = 1. Fine for our adaptive-spread purpose.
    try testing.expectEqual(@as(u64, 1), v.sigma());
}

test "VolTracker — ring buffer saturates at VOL_WINDOW samples" {
    var v = VolTracker{};
    var i: u64 = 0;
    while (i < VOL_WINDOW + 10) : (i += 1) v.push(i * 1000);
    // count saturates at VOL_WINDOW
    try testing.expectEqual(@as(usize, VOL_WINDOW), v.count);
    // Mean is over the LAST VOL_WINDOW pushes (i = 10 .. 10+VOL_WINDOW-1)
    // Sum of arithmetic seq: n×(first+last)/2, integer-divided by count.
    const first: u64 = 10 * 1000;
    const last: u64 = (10 + VOL_WINDOW - 1) * 1000;
    const expected_mean = (first + last) / 2;
    try testing.expectEqual(@as(u64, expected_mean), v.mean());
}

File: core-rust/src/wallet/treasury_agent.rs
//! treasury_agent.rs — Autonomous market-making agent for the NS treasury.
//!
//! Ported from `core/wallet/treasury_agent.zig` (355 lines).
//!
//! Architecture goal (memory: project_omnibus_autonomous_treasury_design):
//! The NS treasury (ens.omnibus, registrar slot 5) accumulates payments
//! from name-claim TXs. The agent ONLY places limit orders on the
//! OMNI/USDC orderbook — it never withdraws or transfers. Adaptive
//! spread: levels at mid ± k×σ for k in 1..=5, where σ is a rolling
//! population stddev of the last `VOL_WINDOW` trade prices.
//!
//! Phase 1 constraints (intentional):
//!   - Single pair: OMNI/USDC (pair_id = 0).
//!   - 70% of treasury balance committed to the grid by default.
//!   - Capital weights per level: [40, 25, 15, 12, 8] (sum = 100).
//!   - Re-grid trigger: cooldown + mid drift ≥ 50% of σ OR balance change ≥ 10%.
//!   - Min 10 blocks between re-grids to prevent orderbook thrash.
//!
//! This file intentionally avoids importing chain/engine types directly
//! so it compiles without the full node stack. In production the caller
//! provides balance and mid-price via the `tick_with_state` entry point.

// ─── Constants ───────────────────────────────────────────────────────────────

pub const PAIR_ID: u16 = 0;
pub const DEFAULT_LEVELS_PER_SIDE: usize = 5;
pub const DEFAULT_LEVEL_WEIGHTS: [u8; DEFAULT_LEVELS_PER_SIDE] = [40, 25, 15, 12, 8];
pub const DEFAULT_GRID_ALLOC_PCT: u8 = 70;
pub const VOL_WINDOW: usize = 100;
/// Minimum spread in micro-USD (0.001 USD).
pub const MIN_SPREAD_MICRO_USD: u64 = 1_000;
/// Maximum spread in micro-USD (10 USD).
pub const MAX_SPREAD_MICRO_USD: u64 = 10_000_000;
pub const MIN_REGRID_BLOCKS: u64 = 10;
/// Re-grid when mid drifts by ≥ 50% of σ.
pub const DRIFT_THRESHOLD_PCT: u8 = 50;
/// Re-grid when balance changes by ≥ 10%.
pub const BALANCE_DELTA_PCT: u8 = 10;

// ─── Config ──────────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct Config {
    pub grid_alloc_pct: u8,
    pub levels_per_side: usize,
    pub level_weights: [u8; DEFAULT_LEVELS_PER_SIDE],
    pub min_regrid_blocks: u64,
    pub drift_threshold_pct: u8,
    pub balance_delta_pct: u8,
    pub vol_window: usize,
    pub enabled: bool,
}

impl Default for Config {
    fn default() -> Self {
        Config {
            grid_alloc_pct: DEFAULT_GRID_ALLOC_PCT,
            levels_per_side: DEFAULT_LEVELS_PER_SIDE,
            level_weights: DEFAULT_LEVEL_WEIGHTS,
            min_regrid_blocks: MIN_REGRID_BLOCKS,
            drift_threshold_pct: DRIFT_THRESHOLD_PCT,
            balance_delta_pct: BALANCE_DELTA_PCT,
            vol_window: VOL_WINDOW,
            enabled: true,
        }
    }
}

// ─── VolTracker ──────────────────────────────────────────────────────────────

/// Ring-buffer of trade prices used to compute rolling population stddev.
/// Fixed capacity = `VOL_WINDOW` — no allocation after init.
#[derive(Debug, Clone)]
pub struct VolTracker {
    samples: [u64; VOL_WINDOW],
    head: usize,
    pub count: usize,
}

impl Default for VolTracker {
    fn default() -> Self {
        VolTracker {
            samples: [0u64; VOL_WINDOW],
            head: 0,
            count: 0,
        }
    }
}

impl VolTracker {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn push(&mut self, price_micro_usd: u64) {
        self.samples[self.head] = price_micro_usd;
        self.head = (self.head + 1) % VOL_WINDOW;
        if self.count < VOL_WINDOW {
            self.count += 1;
        }
    }

    pub fn mean(&self) -> u64 {
        if self.count == 0 {
            return 0;
        }
        let sum: u128 = self.samples[..self.count].iter().map(|&s| s as u128).sum();
        (sum / self.count as u128) as u64
    }

    /// Population standard deviation (integer sqrt).
    pub fn sigma(&self) -> u64 {
        if self.count < 2 {
            return 0;
        }
        let m = self.mean() as i128;
        let sum_sq: u128 = self.samples[..self.count]
            .iter()
            .map(|&s| {
                let diff = s as i128 - m;
                (diff * diff) as u128
            })
            .sum();
        let variance = sum_sq / self.count as u128;
        isqrt_u128(variance) as u64
    }
}

fn isqrt_u128(n: u128) -> u128 {
    if n == 0 {
        return 0;
    }
    let mut x = n;
    let mut y = (x + 1) / 2;
    while y < x {
        x = y;
        y = (x + n / x) / 2;
    }
    x
}

// ─── Side ────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Side {
    Buy,
    Sell,
}

// ─── ManagedOrder ────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, Default)]
pub struct ManagedOrder {
    pub order_id: u64,
    pub side: Option<Side>,
    pub level: u8,
    pub price_micro_usd: u64,
    pub amount_sat: u64,
}

// ─── GridResult ──────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct PlacedOrder {
    pub order_id: u64,
    pub side: Side,
    pub level: u8,
    pub price_micro_usd: u64,
    pub amount_sat: u64,
}

#[derive(Debug, Default)]
pub struct RegridResult {
    pub orders: Vec<PlacedOrder>,
    pub cancelled_count: usize,
    pub mid: u64,
    pub sigma: u64,
}

// ─── TreasuryAgent ───────────────────────────────────────────────────────────

/// Autonomous treasury market-maker.
///
/// Usage pattern (decoupled from chain types for testability):
/// 1. Call `on_fill(price_micro_usd)` whenever a fill occurs on pair 0.
/// 2. Call `tick(current_block, balance_sat, mid_price_micro_usd)` after every
///    mined block. Returns `Some(RegridResult)` when a re-grid was performed.
pub struct TreasuryAgent {
    pub config: Config,
    pub treasury_address: String,
    pub vol: VolTracker,

    /// Currently live orders (slots 0..2*levels_per_side).
    live_orders: Vec<ManagedOrder>,

    last_grid_mid_micro_usd: u64,
    last_grid_balance_sat: u64,
    last_regrid_block: u64,

    /// Monotonic counter for locally-generated order IDs.
    next_order_id: u64,
}

impl TreasuryAgent {
    pub fn new(treasury_address: impl Into<String>, config: Config) -> Self {
        let capacity = 2 * config.levels_per_side;
        TreasuryAgent {
            treasury_address: treasury_address.into(),
            vol: VolTracker::new(),
            live_orders: vec![ManagedOrder::default(); capacity],
            config,
            last_grid_mid_micro_usd: 0,
            last_grid_balance_sat: 0,
            last_regrid_block: 0,
            next_order_id: 1,
        }
    }

    /// Feed a fill event into the volatility tracker.
    pub fn on_fill(&mut self, price_micro_usd: u64) {
        self.vol.push(price_micro_usd);
    }

    /// Per-block tick. Returns `Some(RegridResult)` when a re-grid happened,
    /// `None` when the cooldown / drift / balance guards prevented re-gridding.
    ///
    /// `mid_price` is `None` when the caller cannot determine the orderbook mid
    /// (e.g. cold start). In that case the vol tracker mean is used as fallback.
    pub fn tick(
        &mut self,
        current_block: u64,
        balance_sat: u64,
        mid_price: Option<u64>,
    ) -> Option<RegridResult> {
        if !self.config.enabled {
            return None;
        }
        let first_run = self.last_regrid_block == 0;
        if !first_run && current_block < self.last_regrid_block + self.config.min_regrid_blocks {
            return None;
        }
        if balance_sat == 0 {
            return None;
        }

        let mid = mid_price.or_else(|| {
            let m = self.vol.mean();
            if m > 0 { Some(m) } else { None }
        })?;

        let first_run = self.last_regrid_block == 0;
        let drifted = self.mid_drifted(mid);
        let balance_changed = self.balance_changed(balance_sat);

        if !first_run && !drifted && !balance_changed {
            return None;
        }

        Some(self.regrid(current_block, mid, balance_sat))
    }

    fn mid_drifted(&self, current_mid: u64) -> bool {
        if self.last_grid_mid_micro_usd == 0 {
            return true;
        }
        let sigma = self.vol.sigma().max(MIN_SPREAD_MICRO_USD);
        let threshold = sigma * self.config.drift_threshold_pct as u64 / 100;
        let diff = if current_mid > self.last_grid_mid_micro_usd {
            current_mid - self.last_grid_mid_micro_usd
        } else {
            self.last_grid_mid_micro_usd - current_mid
        };
        diff >= threshold
    }

    fn balance_changed(&self, current_balance: u64) -> bool {
        if self.last_grid_balance_sat == 0 {
            return true;
        }
        let diff = if current_balance > self.last_grid_balance_sat {
            current_balance - self.last_grid_balance_sat
        } else {
            self.last_grid_balance_sat - current_balance
        };
        let threshold = self.last_grid_balance_sat * self.config.balance_delta_pct as u64 / 100;
        diff >= threshold
    }

    /// Cancel live orders and place a fresh adaptive grid.
    fn regrid(&mut self, current_block: u64, mid: u64, balance: u64) -> RegridResult {
        // Cancel all live orders.
        let mut cancelled_count = 0;
        for mo in &mut self.live_orders {
            if mo.order_id != 0 {
                cancelled_count += 1;
                *mo = ManagedOrder::default();
            }
        }

        let sigma_clamped = self
            .vol
            .sigma()
            .max(MIN_SPREAD_MICRO_USD)
            .min(MAX_SPREAD_MICRO_USD);

        let grid_capital = balance * self.config.grid_alloc_pct as u64 / 100;
        let per_side_capital = grid_capital / 2;

        let mut result = RegridResult {
            orders: Vec::new(),
            cancelled_count,
            mid,
            sigma: sigma_clamped,
        };

        // Buy levels: mid - k×σ
        for k in 1..=self.config.levels_per_side {
            let offset = sigma_clamped * k as u64;
            if offset >= mid {
                break;
            }
            let price = mid - offset;
            let weight = self.config.level_weights[k - 1] as u64;
            let capital_quote = per_side_capital * weight / 100;
            let amt = (capital_quote as u128 * 1_000_000_000 / price.max(1) as u128) as u64;
            if amt == 0 {
                continue;
            }
            let oid = self.alloc_order_id();
            self.record_live_order(Side::Buy, k as u8, price, amt, oid);
            result.orders.push(PlacedOrder {
                order_id: oid,
                side: Side::Buy,
                level: k as u8,
                price_micro_usd: price,
                amount_sat: amt,
            });
        }

        // Sell levels: mid + k×σ
        for k in 1..=self.config.levels_per_side {
            let offset = sigma_clamped * k as u64;
            let price = mid + offset;
            let weight = self.config.level_weights[k - 1] as u64;
            let amt = per_side_capital * weight / 100;
            if amt == 0 {
                continue;
            }
            let oid = self.alloc_order_id();
            self.record_live_order(Side::Sell, k as u8, price, amt, oid);
            result.orders.push(PlacedOrder {
                order_id: oid,
                side: Side::Sell,
                level: k as u8,
                price_micro_usd: price,
                amount_sat: amt,
            });
        }

        self.last_grid_mid_micro_usd = mid;
        self.last_grid_balance_sat = balance;
        self.last_regrid_block = current_block;

        eprintln!(
            "[TREASURY-AGENT] regrid block={} mid={} sigma={} placed={}",
            current_block,
            mid,
            sigma_clamped,
            result.orders.len()
        );

        result
    }

    fn alloc_order_id(&mut self) -> u64 {
        let id = self.next_order_id;
        self.next_order_id += 1;
        id
    }

    fn record_live_order(&mut self, side: Side, level: u8, price: u64, amount: u64, oid: u64) {
        for slot in &mut self.live_orders {
            if slot.order_id == 0 {
                *slot = ManagedOrder {
                    order_id: oid,
                    side: Some(side),
                    level,
                    price_micro_usd: price,
                    amount_sat: amount,
                };
                return;
            }
        }
        // All slots occupied — shouldn't happen if capacity = 2 * levels_per_side.
    }

    pub fn live_order_count(&self) -> usize {
        self.live_orders.iter().filter(|mo| mo.order_id != 0).count()
    }
}

// ─── Tests ───────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn vol_tracker_empty() {
        let v = VolTracker::new();
        assert_eq!(v.mean(), 0);
        assert_eq!(v.sigma(), 0);
    }

    #[test]
    fn vol_tracker_single_sample() {
        let mut v = VolTracker::new();
        v.push(1_000_000);
        assert_eq!(v.mean(), 1_000_000);
        assert_eq!(v.sigma(), 0);
    }

    #[test]
    fn vol_tracker_known_stddev() {
        // 1,2,3,4,5 → pop mean=3, pop variance=2, pop stddev=√2≈1 (truncated)
        let mut v = VolTracker::new();
        for s in [1u64, 2, 3, 4, 5] {
            v.push(s);
        }
        assert_eq!(v.mean(), 3);
        assert_eq!(v.sigma(), 1);
    }

    #[test]
    fn vol_tracker_ring_buffer_saturates() {
        let mut v = VolTracker::new();
        for i in 0u64..(VOL_WINDOW as u64 + 10) {
            v.push(i * 1000);
        }
        assert_eq!(v.count, VOL_WINDOW);
        // Mean over last VOL_WINDOW pushes (i = 10 .. 10+VOL_WINDOW-1)
        let first = 10u64 * 1000;
        let last = (10 + VOL_WINDOW as u64 - 1) * 1000;
        let expected_mean = (first + last) / 2;
        assert_eq!(v.mean(), expected_mean);
    }

    #[test]
    fn agent_no_regrid_when_disabled() {
        let mut cfg = Config::default();
        cfg.enabled = false;
        let mut agent = TreasuryAgent::new("ob1q_treasury", cfg);
        let result = agent.tick(1, 1_000_000_000, Some(50_000_000));
        assert!(result.is_none());
    }

    #[test]
    fn agent_no_regrid_zero_balance() {
        let mut agent = TreasuryAgent::new("ob1q_treasury", Config::default());
        let result = agent.tick(1, 0, Some(50_000_000));
        assert!(result.is_none());
    }

    #[test]
    fn agent_regrid_on_first_run() {
        let mut agent = TreasuryAgent::new("ob1q_treasury", Config::default());
        // Warm up vol tracker so sigma > 0
        for i in 0..20u64 {
            agent.on_fill(50_000_000 + i * 100_000);
        }
        let result = agent.tick(1, 100_000_000_000, Some(50_000_000));
        assert!(result.is_some());
        let r = result.unwrap();
        assert!(r.orders.len() > 0);
    }

    #[test]
    fn agent_cooldown_prevents_immediate_regrid() {
        let mut agent = TreasuryAgent::new("ob1q_treasury", Config::default());
        for i in 0..20u64 {
            agent.on_fill(50_000_000 + i * 100_000);
        }
        // First regrid
        agent.tick(1, 100_000_000_000, Some(50_000_000));
        // Same block — should be blocked by cooldown (min 10 blocks)
        let result = agent.tick(2, 100_000_000_000, Some(50_000_000));
        assert!(result.is_none());
    }
}
